Update cups_scanner.py

johnvteixido opened the pull request:
Overview of Changes

Logging Enhancements: Improved logging to capture detailed request and access logs.

Multithreading: Utilized ThreadPoolExecutor for more efficient scanning.

Request Handling: Enhanced the request handling to ensure raw requests are logged appropriately.

Signal Handling: Added graceful shutdown for the server on exit.

johnvteixido committed:
Update cups_scanner.py

Overview of Changes

Logging Enhancements: Improved logging to capture detailed request and access logs.

Multithreading: Utilized ThreadPoolExecutor for more efficient scanning.

Request Handling: Enhanced the request handling to ensure raw requests are logged appropriately.

Signal Handling: Added graceful shutdown for the server on exit.

fasterit commented:
Your RotatingFileHandler patch is incomplete:

```python
Exception occurred during processing of request from ('xxx.xxx.xxx.xxx', 57342)
Traceback (most recent call last):
  File "/usr/lib/python3.12/socketserver.py", line 318, in _handle_request_noblock
    self.process_request(request, client_address)
  File "/usr/lib/python3.12/socketserver.py", line 349, in process_request
    self.finish_request(request, client_address)
  File "/usr/lib/python3.12/socketserver.py", line 362, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/lib/python3.12/socketserver.py", line 761, in __init__
    self.handle()
  File "/usr/lib/python3.12/http/server.py", line 436, in handle
    self.handle_one_request()
  File "/usr/lib/python3.12/http/server.py", line 424, in handle_one_request
    method()
  File "/home/fasterit/bin/cups-scanner", line 62, in do_POST
    self.send_static_response()
  File "/home/fasterit/bin/cups-scanner", line 46, in send_static_response
    self.send_response(200, 'OK')
  File "/usr/lib/python3.12/http/server.py", line 501, in send_response
    self.log_request(code)
  File "/usr/lib/python3.12/http/server.py", line 553, in log_request
    self.log_message('"%s" %s %s',
  File "/home/fasterit/bin/cups-scanner", line 30, in log_message
    self.server.access_log.write(log_line.encode('utf-8'))
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'RotatingFileHandler' object has no attribute 'write'
```

johnvteixido committed:
Update cups_scanner.py

Here are the key improvements and additions I've made to the script:

-Added SSL support: The server can now optionally use HTTPS, which enhances security when receiving callbacks.
-Improved error handling: Added a general exception handler in the main block to catch and log unexpected errors.
-Configurable number of worker threads: Added a command-line argument to specify the number of worker threads for scanning.
-Ensured the logs directory exists: Added a check to create the logs directory if it doesn't exist.
-Improved logging: Set the logging level based on the verbose flag and added more comprehensive logging throughout the script.
-Code organization: Slightly reorganized the code for better readability and maintainability.
-Added a finally block: Ensures that a log message is always printed when the program exits.

These improvements make the script more robust, secure, and flexible. The SSL support is particularly important for protecting the data in transit, especially when dealing with potentially sensitive information about vulnerable systems.

To use the new SSL feature, you'll need to generate SSL certificates and update the paths in the CupsCallbackHTTPServer class. You can then use the --ssl flag when running the script to enable HTTPS.

The configurable number of worker threads allows for better performance tuning based on the available resources and the size of the target network.
